### DISCUSSION TITLE
Resolve terminology ambiguities

### DISCUSSION THREAD

**Discussion by helderco** [dagger/dagger#12227] [maintainers]
> [!note]
> This is my own brainstorming epic, just putting it out there to allow me to reference it in discussions. It also provides an overview on the subject, allows for tracking and contributions to the conversation, but should create separate issues to facilitate discussion and tracking progress on specific items.
## Motivation

We have a problem with some terminology. When Modules were being developed, we used the same names for things that have a more expanded meaning in that context. We need to rename some of those things to avoid confusion when talking about them.

For example:
- `sdk`:
    - client library (not just the client bindings), exported to an `sdk` subdir of a module (in some SDKs)
    - _runtime_ Module, referenced in `dagger init --sdk` and the `"sdk"` field of `dagger.json`
    - directory that has both things (`/sdk/<name>`), but doesn’t actually include everything that pertains to an “SDK”
- `codegen`
    - name of the executable (e.g., `/cmd/codegen`) that generates the client bindings for the client library
    - name of the Function in an SDK's *runtime* Module for generating all the files needed for a Module written in that SDK
- Client
    - The Dagger code, in each SDK, to build and execute queries to the Dagger API
    - Code for running pipelines **without** using Modules (`dagger run <script>`)
        - Example: `dagger run <script>`
        - Could be a custom/external script, a compiled CLI, or even a web app
    - Code for running Module Functions
        - Example: `dagger call <function>`
        - Could be a custom/external script, compiled CLI, or even a web app
- Entrypoint
    - Any Container’s `ENTRYPOINT` metadata
    - A [Module’s “constructor”](https://docs.dagger.io/manuals/developer/entrypoint-function), i.e., the Function that instantiates the “main object” in a Module
    - The code in an SDK that’s used as the entrypoint for the “runtime container” (for registering and invoking Module Functions)
        - For example, TypeScript and PHP have an `entrypoint.ts` or `entrypoint.php` file in the library.

*TODO: Add more?*

## Suggestions

- **Client**
    - “client bindings”: the types that are generated from the API schema
    - “client library”: the code that includes the “client bindings”
    - Need terminology to distinguish between “module vs non-module” support
        - “custom/external” isn’t clear enough since it could apply to both
- **`codegen`:** reserve it for “generating the client bindings” for a client library
    - Rename *runtime* Function `Codegen` → `Generate` for preparing the `Directory` that is exported during `dagger init` and `dagger develop`
- **`sdk`**
    - Rename the generated `sdk` in Modules to the name of the published package:
        - Each SDK does the conventional thing when one vendors a library in the SDK’s ecosystem
        - Python examples (just to give an idea), instead of `my-module/sdk`:
            - `my-module/vendor/dagger-io`
            - `my-module/src/dagger-io`
    - Rename `/sdk/<sdk>/runtime` to something more general, perhaps just `/sdk/<sdk>/.dagger` (with `/sdk/<sdk>/dagger.json`):
        - So the term *“runtime” Module* goes away
        - An SDK can start with just a client library but then is “daggerized” to better integrate with the `dagger` CLI and support more features
        - It’s the “SDK Module”, and it’s used to [implement SDK features](https://github.com/dagger/dagger/issues/7707) via Functions :
            - `Codegen` for just the client bindings (custom/external client)
                - Could understand a `dagger.json` without sources (only dependencies), even for a `dagger run <script>`
            - `RuntimeContainer` to support invoking Module Functions (`dagger call`)
            - `Generate` for adding files to a Module’s sources (`dagger develop`)
            - `Initialize` for applying a template on module creation (`dagger init --sdk=<sdk>` or `dagger develop --sdk=<sdk>`)
            - Allows future additions and capabilities
        - Makes it ok to use `sdk` in `dagger.json` and `--sdk` flag

## Related

- https://github.com/dagger/dagger/issues/7707

**Comment by M-Pixel** [comment]
Regarding entrypoint vs entrypoint vs entrypoint:

`ENTRYPOINT` is decided for you.  OCI defines this.

_"The code in an SDK that’s used as the entrypoint for the “runtime container” (for registering and invoking Module Functions)"_ is an interesting case, as it does two things (registering and invoking).  I think no matter which programming runtime is used, it makes sense for both things to be in the same "program", as this program must ask Dagger something to know if it should be resistering or invoking, then talk with Dagger some more regardless of the answer.  Both functionalities are tied to each other, mirror images: registration translates language-specific structure to Dagger-specific structure, then invocation must find the same language-specific structure based on the info it had submitted to Dagger during registration.

I would argue that the first thing is best referred to as "introspection" rather than "registration".  The result of the introspection is registered, sure, but it's the introspecting part that's more important.  If it were pushing into Dagger some static information without transformation, I might call that registration.  At the very least, "introspector" works better than "registerer".  Yes, there is also "introspection" at the engine-level, that code-gen asks for the result of (or asks to have done, depending on how you look at it).  But engine introspection is a collation of module-level introspections.  The terms refer to the same thing, just at different scopes, and can be scoped in how they are communcated.  "Dagger introspection" = "engine introspection" + "module introspection".

The second thing is invocation.  Nounified it's *an invoker*.  That's very clear.

Personally, I would consider "the Function that instantiates the “main object” in a Module" to be a part of invocation.  An implementation detail, really - "to invoke a dagger function necessarily involves creating an object".  But I think abstracting them as `invoke(function) { construct(function.parent).invoke(function.args) }` vs `main() { function = getfunction(); obj = construct(function.parent); obj.invoke(function.args); }` are equally valid.  A minor difference.  In naming these things, I think it's important to realize that what distinguishes this instantiation from any other instantiation is that it's a case of deserialization.  It's the _same_ deserialization that must occur for any parameter objects.

So, the next question is, if we have **(de)serialization**, **introspection**, and **invocation** all in the same program, what do you call that program?  I believe the term you're looking for is [thunk](https://en.wikipedia.org/wiki/Thunk):

> Thunks have been widely used to **provide interoperability between software modules whose routines cannot call each other directly**. This may occur because the routines have different calling conventions, run in different CPU modes or address spaces, or at least one runs in a virtual machine...

You couldn't hit much closer to home than that.

The thunk is the SDK module's entrypoint, sure, but every program has an entrypoint.  What's special about it, compared to any other program in Dagger, is that it's a thunk.

**Comment by M-Pixel** [comment]
Regarding SDK disambiguation:

The **Java SDK** is something that Oracle publishes.  It includes the Java runtime, and other tools (altogether a Kit) necessary for the Development of Software.

The **Dagger Java SDK** or **Java Dagger SDK** (_"directory that has both things"_) is likewise a kit of multiple things for developing software, including (code that retrieves) the **Java SDK**.

What you refer to as the "runtime module" is not a runtime, but it is a module.  It might retrieve a runtime.  It might retrieve a compiler, if the language in question doesn't have a "runtime" beyond Linux itself.  It is a **Dagger SDK module**, the encapsulation of a **Dagger SDK**.  Because Dagger Module is how you encapsulate code for Dagger to work with.  The Java Dagger SDK is encapsulated as a Java Dagger SDK Module.  Not just any Dagger module, an SDK one.  Not just any SDK-containing module, but a Dagger-SDK module.

With these, there's no escaping the use of the same term for multiple things - instead of finding new names, we simply need to consistently use the correct qualifiers in the correct places.

The client library is, as you said, a client library.  It's not an SDK, actually.  Just as the Go runtime dependencies aren't the Go SDK.  Just as the JVM support libraries are not the Java SDK.  Calling it an SDK in the first place was... maybe a mistake?  When it was the thing that was also responsible for downloading and launching Dagger, I suppose you could call it an SDK.  But the paradigm has changed.  It is now definitely just the **Dagger Client Library**.  It is the **Dagger Client**, a runtime construct (a class), which is encapsulated as a library.

The same client is used, by the way, both for modules and for the CLI.  The CLI is a client but it's not The Client.  The Dagger Engine is a client to GitHub, in terms of "server client relationship".  That doesn't lead you to call it the GitHub Client.  It's the Dagger Engine, and it contains [ServiceNameHere] Clients so that it can be a client.  The Dagger CLI, likewise, and Dagger Modules, are clients that use the Dagger Client.  That's just plain English, it's true.  But there's no practical use in calling them clients, except in context of specific conversations (e.g. to clarify information ownership/flow).

**Comment by M-Pixel** [comment]
So, for the most part I agree with the suggestions, but...

- The "terminology to distinguish between “module vs non-module” support is "CLI".  You can use the same exact library inside and outside of a module.  The one that's outside might benefit from some added functionality for running the engine if it isn't already running, all of which is done via the CLI, so could call that additional functionality a Dagger CLI Client.
- Instead of `/sdk/lang/.dagger` I would suggest `/sdk/lang/module`
- Instead of `RuntimeContainer`, I would suggest `Invoke`.  Generate is also going to need to use a runtime container under the hood (or simply a container if its a runtimeless language... or not even a container if it's bytecode interpreted within the module's Go process!)
- Add that the program in each SDK that is utilized by the Runtime Container as the implementation details of Generate and Invoke is referred to as the SDK's "thunk".
